vulnerabilities/import_runner.py

```diff
@@ -92,54 +156,84 @@ def _process_added_advisories(data_source: DataSource) -> None:
 
-            for id_ in advisory.reference_ids:
-                models.VulnerabilityReference.objects.get_or_create(
-                    vulnerability=vuln, reference_id=id_)
+    else:
+        vuln_refs_qs = models.VulnerabilityReference.objects.filter(
+            vulnerability=vuln)
+        vuln_ids = {ref.reference_id for ref in vuln_refs_qs}
+        vuln_urls = {ref.url for ref in vuln_refs_qs}
 
-            for url in advisory.reference_urls:
-                models.VulnerabilityReference.objects.get_or_create(vulnerability=vuln, url=url)
+        for id_ in advisory.reference_ids:
+            # Add the item preventing duplicates pass to through.
+            if id_ not in vuln_ids:
+                vuln_ids.add(id_)
+                vuln_references.add(VulnerabilityReference_inserter(
+                    vulnerability=vuln, reference_id=id_))
 
-            for ipkg_url in advisory.impacted_package_urls:
-                pkg, created = _get_or_create_package(ipkg_url)
+        for url in advisory.reference_urls:
+            # Add the item preventing duplicates pass to through.
+            if url not in vuln_urls:
+                vuln_urls.add(url)
+                vuln_references.add(VulnerabilityReference_inserter(
+                    vulnerability=vuln, url=url))
 
-                # FIXME Does not work yet due to cascading deletes.
-                # if not created:
-                #     qs = models.ResolvedPackage.objects.filter(
-                #         vulnerability_id=vuln.id, package_id=pkg.id)
-                #     if qs:
-                #         qs[0].delete()
+    return vuln, vuln_created, vuln_references
 
-                models.ImpactedPackage.objects.get_or_create(
-                    vulnerability_id=vuln.id, package_id=pkg.id)
 
-            for rpkg_url in advisory.resolved_package_urls:
-                pkg, created = _get_or_create_package(rpkg_url)
+def _create_pkg_vuln_refs(vuln: models.Vulnerability, vuln_created: bool, purls: Sequence[PackageURL], is_vulnerable: bool):  # nopep8
+    new_refs, updated_refs = set(), set()
+    for purl in purls:
+        pkg, pkg_created = _get_or_create_package(purl)
+        vuln_pkg_ref = PackageRelatedVulnerability_inserter(
+            package=pkg, vulnerability=vuln, is_vulnerable=is_vulnerable)
+        if pkg_created or vuln_created:
+            new_refs.add(vuln_pkg_ref)
+
+        else:
+            qs = models.PackageRelatedVulnerability.objects.filter(
+                package=pkg, vulnerability=vuln)
+            if not qs:
+                new_refs.add(vuln_pkg_ref)
+            else:
+                # Note: PackageRelatedVulnerability has constraints
+                # unique_together = ('package', 'vulnerability', 'is_vulnerable')
+                # This fact is used below.
+                vuln_impact = {i.is_vulnerable for i in qs}
+                if len(vuln_impact) == 2 or is_vulnerable not in vuln_impact:
```

Comment:
What does ==2  mean? that's pretty cryptic

Reply:
This definitely needs an explanation. I have changed  `len(vuln_impact) == 2` to `vuln_impact == {True, False}`, having done that, it still deserves an explanation, I've added the following comments to the code. I hope this clarifies things. 
```
               is_vulnerable is a boolean which indicates the relationship of
                a vulnerability's impact on a package. vuln_impact is a set of
                all such booleans for a pair of (vulnerability, package). In cases
                where  vuln_impact == {True, False}, we know that conflicting relationships
                of (vulnerability, package) ALREADY EXIST in the DB.
                The other check `is_vulnerable not in vuln_impact` is used to know whether the
                data we just found is not conflicting with the data already existing in DB.
                In any of the above two cases we move the entries involved in ImportProblem
```